File: VB/GantSolution.Module.Win/Controllers/RibbonCustomizationWindowController.vb
Imports DevExpress.ExpressApp
Imports DevExpress.XtraBars.Ribbon
Imports DevExpress.XtraGantt
Imports DevExpress.XtraGantt.Ribbon
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks

Namespace GantSolution.Module.Win.Controllers
    Public Class RibbonCustomizationWindowController
        Inherits ViewController(Of ListView)

        Private ganttControl As GanttControl
        Private ribbonControl As RibbonControl
        Private ganttViewRibbonPage As GanttViewRibbonPage
        Public Sub New()
            MyBase.New()
            TargetObjectType = GetType(GanttSolution.Module.BusinessObjects.Task)
        End Sub
        Protected Overrides Sub OnActivated()
            MyBase.OnActivated()
            AddHandler View.ControlsCreated, AddressOf View_ControlsCreated
        End Sub
        Private Sub View_ControlsCreated(ByVal sender As Object, ByVal e As EventArgs)
            Dim ribbonForm As RibbonForm = TryCast(Frame.Template, RibbonForm)
            If ribbonForm IsNot Nothing AndAlso ribbonForm.Ribbon IsNot Nothing Then
                ribbonControl = ribbonForm.Ribbon
                ganttControl = TryCast(View.Editor?.Control, GanttControl)
                If ribbonControl IsNot Nothing AndAlso ganttControl IsNot Nothing Then
                    AddGanttPageToRibbon()
                End If
            End If
        End Sub

        #Region "Initialization"
        Private Sub AddGanttPageToRibbon()
            Dim ribbonStatusBar As New RibbonStatusBar()
            Dim ganttBarController As New GanttBarController()
            Dim ganttSplitViewRibbonPageGroup As New GanttSplitViewRibbonPageGroup()
            ganttViewRibbonPage = New GanttViewRibbonPage()
            Dim ganttAllowResizeBarCheckItem As New GanttAllowResizeBarCheckItem()
            Dim ganttPanelVisibilityDefaultBarCheckItem As New GanttPanelVisibilityDefaultBarCheckItem()
            Dim ganttPanelVisibilityBothBarCheckItem As New GanttPanelVisibilityBothBarCheckItem()
            Dim ganttPanelVisibilityChartBarCheckItem As New GanttPanelVisibilityChartBarCheckItem()
            Dim ganttPanelVisibilityTreeBarCheckItem As New GanttPanelVisibilityTreeBarCheckItem()
            Dim ganttPanelVisibilityBarSubItem As New GanttPanelVisibilityBarSubItem()
            Dim ganttFixedPanelDefaultBarCheckItem As New GanttFixedPanelDefaultBarCheckItem()
            Dim ganttFixedPanelNoneBarCheckItem As New GanttFixedPanelNoneBarCheckItem()
            Dim ganttFixedPanelChartBarCheckItem As New GanttFixedPanelChartBarCheckItem()
            Dim ganttFixedPanelTreeBarCheckItem As New GanttFixedPanelTreeBarCheckItem()
            Dim ganttFixedPanelBarSubItem As New GanttFixedPanelBarSubItem()
            ganttControl.BeginInit()
            ribbonControl.BeginInit()
            CType(ganttBarController, System.ComponentModel.ISupportInitialize).BeginInit()
            ' 
            ' ribbonControl
            ' 
            ribbonControl.Items.AddRange(New DevExpress.XtraBars.BarItem() { ribbonControl.ExpandCollapseItem, ribbonControl.SearchEditItem, ganttAllowResizeBarCheckItem, ganttPanelVisibilityBarSubItem, ganttPanelVisibilityDefaultBarCheckItem, ganttPanelVisibilityBothBarCheckItem, ganttPanelVisibilityChartBarCheckItem, ganttPanelVisibilityTreeBarCheckItem, ganttFixedPanelBarSubItem, ganttFixedPanelDefaultBarCheckItem, ganttFixedPanelNoneBarCheckItem, ganttFixedPanelChartBarCheckItem, ganttFixedPanelTreeBarCheckItem})
            ribbonControl.Pages.AddRange(New RibbonPage() { ganttViewRibbonPage})
            ribbonControl.StatusBar = ribbonStatusBar
            ribbonStatusBar.Ribbon = ribbonControl
            ' 
            ' ganttBarController
            ' 
            ganttBarController.BarItems.Add(ganttAllowResizeBarCheckItem)
            ganttBarController.BarItems.Add(ganttPanelVisibilityDefaultBarCheckItem)
            ganttBarController.BarItems.Add(ganttPanelVisibilityBothBarCheckItem)
            ganttBarController.BarItems.Add(ganttPanelVisibilityChartBarCheckItem)
            ganttBarController.BarItems.Add(ganttPanelVisibilityTreeBarCheckItem)
            ganttBarController.BarItems.Add(ganttPanelVisibilityBarSubItem)
            ganttBarController.BarItems.Add(ganttFixedPanelDefaultBarCheckItem)
            ganttBarController.BarItems.Add(ganttFixedPanelNoneBarCheckItem)
            ganttBarController.BarItems.Add(ganttFixedPanelChartBarCheckItem)
            ganttBarController.BarItems.Add(ganttFixedPanelTreeBarCheckItem)
            ganttBarController.BarItems.Add(ganttFixedPanelBarSubItem)
            ganttBarController.Control = ganttControl
            ' 
            ' ganttSplitViewRibbonPageGroup
            ' 
            ganttSplitViewRibbonPageGroup.CaptionButtonVisible = DevExpress.Utils.DefaultBoolean.False
            ganttSplitViewRibbonPageGroup.ItemLinks.Add(ganttAllowResizeBarCheckItem)
            ganttSplitViewRibbonPageGroup.ItemLinks.Add(ganttPanelVisibilityBarSubItem)
            ganttSplitViewRibbonPageGroup.ItemLinks.Add(ganttFixedPanelBarSubItem)
            ' 
            ' ganttViewRibbonPage
            ' 
            ganttViewRibbonPage.Groups.AddRange(New RibbonPageGroup() { ganttSplitViewRibbonPageGroup})
            ganttViewRibbonPage.Text = "Gantt"
            ' 
            ' ganttPanelVisibilityBarSubItem
            ' 
            ganttPanelVisibilityBarSubItem.LinksPersistInfo.AddRange(New DevExpress.XtraBars.LinkPersistInfo() {
                New DevExpress.XtraBars.LinkPersistInfo(ganttPanelVisibilityDefaultBarCheckItem),
                New DevExpress.XtraBars.LinkPersistInfo(ganttPanelVisibilityBothBarCheckItem),
                New DevExpress.XtraBars.LinkPersistInfo(ganttPanelVisibilityChartBarCheckItem),
                New DevExpress.XtraBars.LinkPersistInfo(ganttPanelVisibilityTreeBarCheckItem)
            })
            ' 
            ' ganttFixedPanelBarSubItem
            ' 
            ganttFixedPanelBarSubItem.LinksPersistInfo.AddRange(New DevExpress.XtraBars.LinkPersistInfo() {
                New DevExpress.XtraBars.LinkPersistInfo(ganttFixedPanelDefaultBarCheckItem),
                New DevExpress.XtraBars.LinkPersistInfo(ganttFixedPanelNoneBarCheckItem),
                New DevExpress.XtraBars.LinkPersistInfo(ganttFixedPanelChartBarCheckItem),
                New DevExpress.XtraBars.LinkPersistInfo(ganttFixedPanelTreeBarCheckItem)
            })
            ganttControl.EndInit()
            ribbonControl.EndInit()
            CType(ganttBarController, System.ComponentModel.ISupportInitialize).EndInit()
        End Sub
        #End Region

        Protected Overrides Sub OnDeactivated()
            RemoveHandler View.ControlsCreated, AddressOf View_ControlsCreated
            If ribbonControl IsNot Nothing Then
                ribbonControl.Pages.Remove(ganttViewRibbonPage)
                ganttViewRibbonPage.Dispose()
            End If
            MyBase.OnDeactivated()
        End Sub
    End Class
End Namespace

File: CS/EF/GanttSolution/GanttSolution.Win/Controllers/RibbonCustomizationWindowController.cs
using DevExpress.ExpressApp;
using DevExpress.XtraBars.Ribbon;
using DevExpress.XtraGantt;
using DevExpress.XtraGantt.Ribbon;
using System;

namespace GanttSolution.Module.Win.Controllers {
    public class RibbonCustomizationWindowController : ViewController<ListView> {
        private GanttControl ganttControl;
        private RibbonControl ribbonControl;
        private GanttViewRibbonPage ganttViewRibbonPage;
        public RibbonCustomizationWindowController() : base() {
            TargetObjectType = typeof(GanttSolution.Module.BusinessObjects.MyTask);
        }
        protected override void OnActivated() {
            base.OnActivated();
            View.ControlsCreated += View_ControlsCreated;
        }
        private void View_ControlsCreated(object sender, EventArgs e) {
            RibbonForm ribbonForm = Frame.Template as RibbonForm;
            if(ribbonForm != null && ribbonForm.Ribbon != null) {
                ribbonControl = ribbonForm.Ribbon;
                ganttControl = View.Editor?.Control as GanttControl;
                if(ribbonControl != null && ganttControl != null) {
                    AddGanttPageToRibbon();
                }
            }  
        }

        #region Initialization
        private void AddGanttPageToRibbon() {
            RibbonStatusBar ribbonStatusBar = new RibbonStatusBar();
            GanttBarController ganttBarController = new GanttBarController();
            GanttSplitViewRibbonPageGroup ganttSplitViewRibbonPageGroup = new GanttSplitViewRibbonPageGroup();
            ganttViewRibbonPage = new GanttViewRibbonPage();
            GanttAllowResizeBarCheckItem ganttAllowResizeBarCheckItem = new GanttAllowResizeBarCheckItem();
            GanttPanelVisibilityDefaultBarCheckItem ganttPanelVisibilityDefaultBarCheckItem = new GanttPanelVisibilityDefaultBarCheckItem();
            GanttPanelVisibilityBothBarCheckItem ganttPanelVisibilityBothBarCheckItem = new GanttPanelVisibilityBothBarCheckItem();
            GanttPanelVisibilityChartBarCheckItem ganttPanelVisibilityChartBarCheckItem = new GanttPanelVisibilityChartBarCheckItem();
            GanttPanelVisibilityTreeBarCheckItem ganttPanelVisibilityTreeBarCheckItem = new GanttPanelVisibilityTreeBarCheckItem();
            GanttPanelVisibilityBarSubItem ganttPanelVisibilityBarSubItem = new GanttPanelVisibilityBarSubItem();
            GanttFixedPanelDefaultBarCheckItem ganttFixedPanelDefaultBarCheckItem = new GanttFixedPanelDefaultBarCheckItem();
            GanttFixedPanelNoneBarCheckItem ganttFixedPanelNoneBarCheckItem = new GanttFixedPanelNoneBarCheckItem();
            GanttFixedPanelChartBarCheckItem ganttFixedPanelChartBarCheckItem = new GanttFixedPanelChartBarCheckItem();
            GanttFixedPanelTreeBarCheckItem ganttFixedPanelTreeBarCheckItem = new GanttFixedPanelTreeBarCheckItem();
            GanttFixedPanelBarSubItem ganttFixedPanelBarSubItem = new GanttFixedPanelBarSubItem();
            ganttControl.BeginInit();
            ribbonControl.BeginInit();
            ((System.ComponentModel.ISupportInitialize)ganttBarController).BeginInit();
            // 
            // ribbonControl
            // 
            ribbonControl.Items.AddRange(new DevExpress.XtraBars.BarItem[] {
            ribbonControl.ExpandCollapseItem,
            ribbonControl.SearchEditItem,
            ganttAllowResizeBarCheckItem,
            ganttPanelVisibilityBarSubItem,
            ganttPanelVisibilityDefaultBarCheckItem,
            ganttPanelVisibilityBothBarCheckItem,
            ganttPanelVisibilityChartBarCheckItem,
            ganttPanelVisibilityTreeBarCheckItem,
            ganttFixedPanelBarSubItem,
            ganttFixedPanelDefaultBarCheckItem,
            ganttFixedPanelNoneBarCheckItem,
            ganttFixedPanelChartBarCheckItem,
            ganttFixedPanelTreeBarCheckItem});
            ribbonControl.Pages.AddRange(new RibbonPage[] {
            ganttViewRibbonPage});
            ribbonControl.StatusBar = ribbonStatusBar;
            ribbonStatusBar.Ribbon = ribbonControl;
            // 
            // ganttBarController
            // 
            ganttBarController.BarItems.Add(ganttAllowResizeBarCheckItem);
            ganttBarController.BarItems.Add(ganttPanelVisibilityDefaultBarCheckItem);
            ganttBarController.BarItems.Add(ganttPanelVisibilityBothBarCheckItem);
            ganttBarController.BarItems.Add(ganttPanelVisibilityChartBarCheckItem);
            ganttBarController.BarItems.Add(ganttPanelVisibilityTreeBarCheckItem);
            ganttBarController.BarItems.Add(ganttPanelVisibilityBarSubItem);
            ganttBarController.BarItems.Add(ganttFixedPanelDefaultBarCheckItem);
            ganttBarController.BarItems.Add(ganttFixedPanelNoneBarCheckItem);
            ganttBarController.BarItems.Add(ganttFixedPanelChartBarCheckItem);
            ganttBarController.BarItems.Add(ganttFixedPanelTreeBarCheckItem);
            ganttBarController.BarItems.Add(ganttFixedPanelBarSubItem);
            ganttBarController.Control = ganttControl;
            // 
            // ganttSplitViewRibbonPageGroup
            // 
            ganttSplitViewRibbonPageGroup.CaptionButtonVisible = DevExpress.Utils.DefaultBoolean.False;
            ganttSplitViewRibbonPageGroup.ItemLinks.Add(ganttAllowResizeBarCheckItem);
            ganttSplitViewRibbonPageGroup.ItemLinks.Add(ganttPanelVisibilityBarSubItem);
            ganttSplitViewRibbonPageGroup.ItemLinks.Add(ganttFixedPanelBarSubItem);
            // 
            // ganttViewRibbonPage
            // 
            ganttViewRibbonPage.Groups.AddRange(new RibbonPageGroup[] {
            ganttSplitViewRibbonPageGroup});
            ganttViewRibbonPage.Text = "Gantt";
            // 
            // ganttPanelVisibilityBarSubItem
            // 
            ganttPanelVisibilityBarSubItem.LinksPersistInfo.AddRange(new DevExpress.XtraBars.LinkPersistInfo[] {
            new DevExpress.XtraBars.LinkPersistInfo(ganttPanelVisibilityDefaultBarCheckItem),
            new DevExpress.XtraBars.LinkPersistInfo(ganttPanelVisibilityBothBarCheckItem),
            new DevExpress.XtraBars.LinkPersistInfo(ganttPanelVisibilityChartBarCheckItem),
            new DevExpress.XtraBars.LinkPersistInfo(ganttPanelVisibilityTreeBarCheckItem)});
            // 
            // ganttFixedPanelBarSubItem
            // 
            ganttFixedPanelBarSubItem.LinksPersistInfo.AddRange(new DevExpress.XtraBars.LinkPersistInfo[] {
            new DevExpress.XtraBars.LinkPersistInfo(ganttFixedPanelDefaultBarCheckItem),
            new DevExpress.XtraBars.LinkPersistInfo(ganttFixedPanelNoneBarCheckItem),
            new DevExpress.XtraBars.LinkPersistInfo(ganttFixedPanelChartBarCheckItem),
            new DevExpress.XtraBars.LinkPersistInfo(ganttFixedPanelTreeBarCheckItem)});
            ganttControl.EndInit();
            ribbonControl.EndInit();
            ((System.ComponentModel.ISupportInitialize)ganttBarController).EndInit();
        }
		#endregion

		protected override void OnDeactivated() {
            View.ControlsCreated -= View_ControlsCreated;
            if(ribbonControl != null) {
                ribbonControl.Pages.Remove(ganttViewRibbonPage);
                ganttViewRibbonPage.Dispose();
            }
            base.OnDeactivated();
        }
    }
}
